pkg/proxy/port_allocator.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proxy
+
+import (
+	"errors"
+	"math/rand"
+	"sync"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+)
+
+var (
+	defaultPortAllocator         = &randomAllocator{}
+	errPortRangeNoPortsRemaining = errors.New("port allocation failed; there are no remaining ports left to allocate in the accepted range")
+)
+
+type PortAllocator interface {
+	AllocateNext() (int, error)
+	Release(int)
+}
+
+// randomAllocator is a PortAllocator implementation that allocates random ports, yielding
+// a port value of 0 for every call to AllocateNext().
+type randomAllocator struct{}
+
+// AllocateNext always returns 0
+func (r *randomAllocator) AllocateNext() (int, error) {
+	return 0, nil
+}
+
+// Release is a noop
+func (r *randomAllocator) Release(_ int) {
+	// noop
+}
+
+// newAllocator builds PortAllocator for a given PortRange. If the PortRange is empty
+// then a random port allocator is returned; otherwise, a new range-based allocator
+// is returned.
+func newAllocator(r util.PortRange) PortAllocator {
+	if r.Base == 0 {
+		return defaultPortAllocator
+	}
+	return newRangeAllocator(r)
+}
+
+type rangeAllocator struct {
+	ports chan uint
+	used  map[uint]struct{}
+	lock  sync.Mutex
+}
+
+func newRangeAllocator(r util.PortRange) PortAllocator {
+	if r.Base == 0 || r.Size == 0 {
+		panic("illegal argument: may not specify an empty port range")
+	}
+	sz := uint(r.Size)
+	ports := make(chan uint, sz)
+
+	// build a queue of shuffled ports
```

Comment:
I guess it's not so terrible to keep a list of ints, but I would still
change the chan to buffer only 1 or 2 or 16 ints and feed it from a
goroutine, keeping state in a bitmap.  Randomly pick a bit in the bitmap,
if not free scan forward until you find a free one, claim it, write that to
the chan.



On Wed, May 20, 2015 at 5:35 AM, James DeFelice <notifications@github.com>
wrote:

> In pkg/proxy/port_allocator.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/8417#discussion_r30695980>
> :
>
> > +}
> > +
> > +type rangeAllocator struct {
> > +	ports chan uint
> > +	used  map[uint]struct{}
> > +	lock  sync.Mutex
> > +}
> > +
> > +func newRangeAllocator(r util.PortRange) PortAllocator {
> > +	if r.Base == 0 || r.Size == 0 {
> > +		panic("illegal argument: may not specify an empty port range")
> > +	}
> > +	sz := uint(r.Size)
> > +	ports := make(chan uint, sz)
> > +
> > +	// build a queue of shuffled ports
>
> Actually, this approach (map+chan) really simplifies allocation (don't
> need random+retry+scan) with a minor storage penalty. I can refactor to use
> a bitset (math/big) to save more space.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/8417/files#r30695980>
> .
>